### PR TITLE
[TechDraw] Microimprove new Face Finder Algorithm

### PR DESCRIPTION
**Pull request by benj5378**
> It is, in fact, somewhat slower than the current algo.

Micro improvements

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [x]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [x]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

### CODE REVIEW

File: src/Mod/TechDraw/App/DrawViewPart.cpp [hunk 1]
```diff
@@ -523,21 +523,21 @@ void DrawViewPart::extractFaces()
...
+                    continue;   //can not make a face from wire with no area
+                }
+
                 TechDraw::FacePtr f(std::make_shared<TechDraw::Face>());
                 const TopoDS_Wire& wire = (*itWire);
-                TechDraw::Wire* w = new TechDraw::Wire(wire);
-                f->wires.push_back(w);
+                f->wires.push_back(new TechDraw::Wire(wire));
                 if (geometryObject) {
                     geometryObject->addFaceGeom(f);
```
Comment (benj5378) [inline src/Mod/TechDraw/App/DrawViewPart.cpp:542]: Could this ever be false?

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7456">pipeline status</a> for feature branch [PR_7456](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7456). Pipeline [633395133](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/633395133) was triggered at [bbe016c](https://github.com/FreeCAD/FreeCAD/pull/7456/commits/bbe016cc1da4abaf82aa83920ab434854e236372). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by WandererFan** [comment]
We used to get segfaults if a new hlr cycle was started while face finding was active.  That's why the check for geometryObject was added.  It is less likely now after a change to retain the old GO until the new one is populated, but still theoretically possible.

**Comment by benj5378** [comment]
> It is less likely now after a change to retain the old GO until the new one is populated, but still theoretically possible.

Also even though we have already checked that `geometryObject` isn't `nullptr` on https://github.com/FreeCAD/FreeCAD/blob/8b4e114b9d381bac271980529164fa3cf33954c6/src/Mod/TechDraw/App/DrawViewPart.cpp#L477-L480

@WandererFan

**Comment by WandererFan** [comment]
What was happening before was that a second hlr thread was starting before
this face finding thread was complete. So even though we checked
geometryObject at the top of the routine, the other thread could modify
geometryObject while extractFaces was processing.   In the current
implementation, geometryObject is only invalid for the execution of two
instructions.

In a perfect world geometryObject would be protected by a mutex to
completely eliminate the possibility of it being invalid.

**Pull request merged**